### PR TITLE
[ONNX] Remove Aten ops from ONNX export

### PR DESCRIPTION
**Pull request by neginraoof**
This PR adds a new operator export type to exporter: ONNX_FALLTHROUGH
This new type allows ops that are not supported to pass through.
This PR also removes all aten ops in ONNX operator export type mode.

### CODE REVIEW

**Comment by vadimkantorov** [comment]
There are a lot of almost duplicate error messages `'Unsupported: ONNX export of cumsum in '
                           'opset 9. Please try higher opset versions.'`. Should a function / format string be put in a variable?

**Comment by neginraoof** [comment]
@vadimkantorov Yeah makes sense. This is mainly WIP now to check CI failures due to removal of ATen ops in ONNX. I'll refactor and clean up export and symbolic conversions a bit.

**Comment by neginraoof** [comment]
@houseroad I'm seeing ONNX checker failures. Would it be possible to update ONNX submodule?

File: test/onnx/expect/TestOperators.test_embedding_bags.expect [hunk 1]
```diff
@@ -10,7 +10,6 @@ graph {
     output: "4"
     output: "5"
     output: "6"
-    name: "ATen_0"
```
Comment (BowenBao) [inline test/onnx/expect/TestOperators.test_embedding_bags.expect:13]: what happened here?..

File: test/onnx/test_utility_funs.py [hunk 2]
```diff
@@ -522,6 +523,44 @@ def forward(self, x):
         # verify that the model state is preserved
         assert model.training == old_state
 
+    # TODO: Enable test when BatchNorm is implemented in ORT for opset 12.
+    @skipIfUnsupportedOpsetVersion([12])
+    def test_batchnorm_training(self):
```
Comment (BowenBao) [inline test/onnx/test_utility_funs.py:528]: is this part of the scope of the PR?

File: torch/onnx/utils.py [hunk 3]
```diff
@@ -729,6 +729,17 @@ def const_if_tensor(arg):
 # help the target of an ONNX export export more efficiently.  However,
 # ONNX doesn't currently formalize inplace. Fortunately, it's sound to drop
 # inplace annotations, but we are losing information this way.
```
Comment (BowenBao) [inline torch/onnx/utils.py:731]: this comment looks unrelated to the below new function. Maybe add more blank lines in between

File: torch/onnx/utils.py [hunk 4]
```diff
@@ -729,6 +729,17 @@ def const_if_tensor(arg):
...
+def _find_symbolic_fn(ns, domain, op_name, opset_version, operator_export_type, msg=''):
+    import torch.onnx.symbolic_registry as sym_registry
+    if not sym_registry.is_registered_op(op_name, domain, opset_version):
+        if operator_export_type == OperatorExportTypes.ONNX_ATEN_FALLTHROUGH:
+            # Use the original node directly
+            return None
+        else:
+            warnings.warn("ONNX export failed on operator {}::{} because "
```
Comment (BowenBao) [inline torch/onnx/utils.py:739]: I'm wondering if we should keep this as warning. What happens for each different operator export type, if code reach this place?

File: torch/onnx/utils.py [hunk 5]
```diff
@@ -807,45 +819,45 @@ def _run_symbolic_function(g, n, inputs, env, operator_export_type=OperatorExpor
...
-                          "Otherwise please report a bug".format(ns, op_name))
+            if operator_export_type != OperatorExportTypes.ONNX_ATEN_FALLTHROUGH:
+                warnings.warn("ONNX export failed on an operator with unrecognized namespace {}::{}; "
+                              "If you are trying to export a custom operator, make sure you registered "
+                              "it with the right domain and version."
+                              "Otherwise please report a bug".format(ns, op_name))
             return None
-
+    except RuntimeError:
+        if operator_export_type == OperatorExportTypes.ONNX_ATEN_FALLTHROUGH:
```
Comment (BowenBao) [inline torch/onnx/utils.py:858]: All runtime errors are hidden for `ONNX_ATEN_FALLTHROUGH`?

File: test/onnx/test_utility_funs.py [hunk 6]
```diff
@@ -543,10 +582,119 @@ def forward(self, x):
...
 
+    def test_aten_fallthrough(self):
+        # Test aten export of op with no symbolic
+        class Module(torch.nn.Module):
+            def forward(self, x):
+                return torch.triu(x)
+
+        x = torch.randn(2, 3, 4)
+        _set_opset_version(self.opset_version)
+        graph, _, __ = utils._model_to_graph(Module(), (x, ),
```
[the remark this answers is not in the document]
Comment (neginraoof) [reply test/onnx/test_utility_funs.py:559]: The purpose is to just check the graph nodes, so we use this method which returns the exported graph.

File: torch/onnx/utils.py (hunk 4 above)
Comment (neginraoof) [reply torch/onnx/utils.py:739]: Yes, this warning just adds more info to the error. But it's duplicate. We can remove this.

File: torch/onnx/utils.py (hunk 5 above)
Comment (neginraoof) [reply torch/onnx/utils.py:858]: All runtime errors that are coming from symbolics are hidden. We will export the op as is, even if the spec is not known to the exporter.

File: test/onnx/expect/TestOperators.test_embedding_bags.expect (hunk 1 above)
Comment (neginraoof) [reply test/onnx/expect/TestOperators.test_embedding_bags.expect:13]: After the update for symbolics to remove ATen ops, test_embedding is only exported with ATEN_FALLBACK.
After this is changes, add_node_names is set to False, since we do not support add_node_names=True for operaotr export types other thatn ONNX.

**Review by houseroad — COMMENTED** [review]
I will try to get it updated this week.

File: torch/onnx/symbolic_opset9.py [hunk 7]
```diff
@@ -307,29 +309,43 @@ def reduce_dim(g, self, dim, keepdim, dtype):
...
 def cumsum(g, input, dim, dtype):
-    if dtype.node().kind() != 'prim::Constant':
-        return _unimplemented(name, "dtype")
-    return g.op("ATen", input, operator_s="cumsum", dim_i=dim)
+    if sym_help._operator_export_type == torch.onnx.OperatorExportTypes.ONNX_ATEN_FALLBACK:
+        if dtype.node().kind() != 'prim::Constant':
+            return _unimplemented(name, "dtype")
+        return g.op("ATen", input, operator_s="cumsum", dim_i=dim)
+    else:
+        raise RuntimeError('Unsupported: ONNX export of cumsum in '
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset9.py:325]: maybe we should abstract out this error message, like `_opset_unsupported('cumsum', current_opset=9, supported_opset=11)`

File: torch/onnx/utils.py [hunk 8]
```diff
@@ -770,16 +781,14 @@ def _run_symbolic_function(g, n, inputs, env, operator_export_type=OperatorExpor
                 outputs = n.outputsSize()
                 attrs["outputs"] = outputs
                 return _graph_at(g, op_name, *inputs, aten=True, **attrs)
-
             else:
                 # Export it regularly
+                domain = ''
+                symbolic_fn = _find_symbolic_in_registry(domain, op_name, opset_version, operator_export_type)
+                if symbolic_fn is None:
+                    return None
```
Comment (BowenBao) [inline torch/onnx/utils.py:789]: what happens for export type 'ONNX' when symbolic_fn is None?

File: torch/onnx/__init__.py [hunk 9]
```diff
@@ -75,10 +75,40 @@ def export(model, args, f, export_params=True, verbose=False, training=TrainingM
...
+            is exported as:
+                graph(%0 : Float):
+                  %1 : Long() = onnx::Constant[value={0}]()
+                  %2 : Float = aten::ATen[operator="triu"](%0, %1)  # missing op
+                  %3 : Float = onnx::Mul(%2, %0) # registered op
+                  return (%3)
+            OperatorExportTypes.RAW: Export raw ir.
+            OperatorExportTypes.ONNX_FALLTHROUGH: If an op is not supported in
+            ONNX, fall through and export the operator as is. This mode can be used
+            if you are trying to export a custom op that is not supported in ONNX.
```
[the remark this answers is not in the document]
Comment (neginraoof) [reply torch/onnx/__init__.py:100]: This is the case where you don't register your op as an onnx op, and the backend can process it.

File: torch/onnx/utils.py (hunk 8 above)
Comment (neginraoof) [reply torch/onnx/utils.py:789]: Symbolic cannot be None for ONNX export type. Because of line 742:

if operator_export_type == OperatorExportTypes.ONNX_FALLTHROUGH:
            # Use the original node directly
            return None

With ONNX export type, sym_registry.get_registered_op(op_name, domain, opset_version) will always be called,
which would through RuntimeError for unregistered ONNX ops.

File: torch/onnx/__init__.py [hunk 10]
```diff
@@ -75,10 +75,40 @@ def export(model, args, f, export_params=True, verbose=False, training=TrainingM
...
+            Example graph:
+                graph(%0 : Float):
+                  %3 : int = prim::Constant[value=0]()
+                  %4 : Float = aten::triu(%0, %3) # missing op
+                  %5 : Float = aten::mul(%4, %0) # registered op
+                  return (%5)
+            is exported as:
+                graph(%0 : Float):
+                  %1 : Long() = onnx::Constant[value={0}]()
+                  %2 : Float = aten::ATen[operator="triu"](%0, %1)  # missing op
```
[the remark this answers is not in the document]
Comment (neginraoof) [reply torch/onnx/__init__.py:94]: FO ONNX_ATEN_FALLBACK, It actually could be both.
If the symbolic does not exist at all, it'll show up as aten:ATen (like triu)
If the op is registered, symbolic exists and returns and ATen node, it'll show up as onnx:ATen (like embedding_bag)

File: torch/onnx/__init__.py [hunk 11]
```diff
@@ -75,10 +75,41 @@ def export(model, args, f, export_params=True, verbose=False, training=TrainingM
...
+                  %3 : Float = onnx::Mul(%2, %0) # registered op
+                  return (%3)
+            OperatorExportTypes.RAW: Export raw ir.
+            OperatorExportTypes.ONNX_FALLTHROUGH: If an op is not supported in
+            ONNX, fall through and export the operator as is. Using this mode,
+            a custom operator can be registered and implemented by the user for
```
Comment (neginraoof) [inline torch/onnx/__init__.py:100]: @BowenBao The description here is updated. Please take a look and let me know if you have a better suggestion. Thanks.

File: torch/onnx/__init__.py [hunk 12]
```diff
@@ -75,10 +75,41 @@ def export(model, args, f, export_params=True, verbose=False, training=TrainingM
...
+            their runtime backend.
+            Example graph:
+                graph(%x.1 : Float):
+                  %1 : Tensor = prim::Uninitialized()
```
Comment (BowenBao) [inline torch/onnx/__init__.py:104]: `%1` in example is unused. maybe switch to a better example

File: torch/onnx/__init__.py (hunk 11 above)
Comment (BowenBao) [reply torch/onnx/__init__.py:100]: I think i mis-understood the `custom op` definition in the description previously. 

"If an op is not exportable to ONNX, fall through and export the operator as is, as a custom ONNX op. Using this mode, it can be registered and implemented by the user for their runtime backend."

**Review by BowenBao — APPROVED** [review]

**Comment by neginraoof** [comment]
cc @houseroad 
Please review this PR. Thanks!

**Comment by neginraoof** [comment]
@houseroad Thanks for looking into this. Could you please give me more details about the internal failures?

**Review by houseroad — COMMENTED** [review]
Could you rebase the PR to resolve the conflict?

File: torch/onnx/__init__.py [hunk 13]
```diff
@@ -75,10 +75,42 @@ def export(model, args, f, export_params=True, verbose=False, training=TrainingM
...
+            Example graph:
+                graph(%x.1 : Long(1:1)):
+                  %1 : None = prim::Constant()
+                  %2 : Tensor = aten::sum(%x.1, %1)
+                  %y.1 : Tensor[] = prim::ListConstruct(%2)
+                  return (%y.1)
+            is exported as:
+                graph(%x.1 : Long(1:1)):
+                  %1 : Tensor = onnx::ReduceSum[keepdims=0](%x.1)
+                  %y.1 : Long() = prim::ListConstruct(%1)
```
Comment (houseroad) [inline torch/onnx/__init__.py:111]: After reading the comments, I am still not clear how ONNX_FALLTHROUGH works, could you also provide an ONNX graph as well?

Comment (neginraoof) [reply torch/onnx/__init__.py:111]: The example graph is at line 103:

Example graph:
                graph(%x.1 : Long(1:1)):
                  %1 : None = prim::Constant()
                  %2 : Tensor = aten::sum(%x.1, %1)
                  %y.1 : Tensor[] = prim::ListConstruct(%2)
                  return (%y.1)
            is exported as:
                graph(%x.1 : Long(1:1)):
                  %1 : Tensor = onnx::ReduceSum[keepdims=0](%x.1)
                  %y.1 : Long() = prim::ListConstruct(%1)
                return (%y.1)

This mode is converting the registered onnx ops into ONNX domain, and fallthrough the ops not registered.

Comment (neginraoof) [reply torch/onnx/__init__.py:111]: cc @houseroad

Comment (houseroad) [reply torch/onnx/__init__.py:111]: Yeah, I think you should add the above explanation to the comments as well, such as in the example, which op is not supported by onnx, and how it falls through, and the major difference between ATEN_FALLBACK and ONNX_FALLTHROUGH

Comment (neginraoof) [reply torch/onnx/__init__.py:111]: Sure, I'll update it.

**Review by houseroad — APPROVED** [review]
